packages/x-data-grid/src/components/grid/toolbar/GridToolbarToggleButton.tsx

```diff
@@ -0,0 +1,116 @@
+import clsx from 'clsx';
+import { useGridRootProps } from '../../../hooks/utils/useGridRootProps';
+import { getDataGridUtilityClass } from '../../../constants/gridClasses';
+import type { DataGridProcessedProps } from '../../../models/props/DataGridProps';
+import {
+  useGridComponentRenderer,
+  RenderProp,
+} from '../../../hooks/utils/useGridComponentRenderer';
+
+export interface GridToolbarToggleButtonProps extends ToggleButtonProps {
+  /**
+   * A function to customize rendering of the component.
+   */
+  render?: RenderProp<{}>;
+}
+
+type OwnerState = DataGridProcessedProps;
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    root: ['toolbarToggleButton'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+const StyledGridToolbarToggleButton = styled(ToggleButton, {
+  name: 'MuiDataGrid',
+  slot: 'ToolbarToggleButton',
+  overridesResolver: (_, styles) => styles.toolbarToggleButton,
+})<{ ownerState: OwnerState }>(({ theme }) => ({
+  gap: theme.spacing(0.5),
+  minHeight: 34,
+  lineHeight: 'normal',
+}));
+
+const DefaultGridToolbarToggleButton = React.forwardRef<
+  HTMLButtonElement,
+  GridToolbarToggleButtonProps
+>(function DefaultGridToolbarToggleButton(props, ref) {
+  const { children, className, size = 'small', ...other } = props;
+  const rootProps = useGridRootProps();
+  const classes = useUtilityClasses(rootProps);
+
+  return (
+    <StyledGridToolbarToggleButton
+      ref={ref}
+      as={rootProps.slots.baseToggleButton}
+      ownerState={rootProps}
+      className={clsx(classes.root, className)}
+      size={size}
+      {...rootProps.slotProps?.baseToggleButton}
+      {...other}
+    >
+      {children}
+    </StyledGridToolbarToggleButton>
+  );
+});
+
+DefaultGridToolbarToggleButton.propTypes = {
+  // ----------------------------- Warning --------------------------------
+  // | These PropTypes are generated from the TypeScript type definitions |
+  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
+  // ----------------------------------------------------------------------
+  /**
+   * A function to customize rendering of the component.
+   */
+  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
+} as any;
+
+/**
+ * Demos:
+ *
+ * - [Toolbar](https://mui.com/x/react-data-grid/components/toolbar/)
+ *
+ * API:
+ *
+ * - [GridToolbarToggleButton API](https://mui.com/x/api/data-grid/grid-toolbar-toggle-button/)
+ */
+const GridToolbarToggleButton = React.forwardRef<HTMLButtonElement, GridToolbarToggleButtonProps>(
+  function GridToolbarToggleButton(props, ref) {
+    const { render, ...other } = props;
+
+    const { renderElement } = useGridComponentRenderer({
+      render,
+      defaultElement: DefaultGridToolbarToggleButton as React.ComponentType<any>,
+      props: {
+        ref,
+        ...other,
+      },
+    });
+
+    return renderElement();
```

Comment:
We're `...` twice in the body of this function, plus the params object for `useGridComponentRenderer`. If we consider only this use-case, the perf cost could be acceptable, but at scale (if many components use this pattern) it would cause perf issues. Is there a way to limit our use of `...`? Could we avoid the params object for `useGridComponentRenderer` and use normal parameters instead?

Reply:
> We're using ... twice

Add to that the `...` uses in the default component implementation above.